### PR TITLE
fix(collector): wait until run pod has been deleted

### LINKED ISSUE #1186: e2e tests fail first time round with an error from runPod
**Bug Description**

On new changes running tests, the e2e test fails with the error:

```
failed to run collector: run/static-hi: failed to run pod: failed to create pod: object is being deleted: pods "static-hi" already exists. 
```

**Expected Behavior**

The pod should be deleted after success, not just after failure.

**Steps To Reproduce**

Make a small change, watch the tests.

However, you could likely also produce a reproducer with a support bundle spec that includes runPod, and watch to see if the pod is cleaned up after the collector finishes.

**Additional Context**

Include the following information.
- Troubleshoot version. If you built from source, note that including the version of Go you used to build with.
- Operating system
- Operating system version
- Other details that might be helpful in diagnosing the problem

Most recent change in that part of the code: https://github.com/replicatedhq/troubleshoot/pull/1172 (might be irrelevant)

### PR DESCRIPTION
**Pull request by DexterYan**
## Description, Motivation and Context

- Make sure the run pod has been deleted before the troubleshoot.sh finished running
- Poll every second to check if the Pod has been deleted. The maximum grace period is 2 Minutes.

Fixes: #1186
Previous PR: #1172 

## Checklist

- [ ] New and existing tests pass locally with introduced changes.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] The commit message(s) are informative and highlight any breaking changes
- [ ] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by DexterYan** [comment]
## Example Support Bundle
```
apiVersion: troubleshoot.sh/v1beta2
kind: SupportBundle
metadata:
  name: example
spec:
  collectors:
    - runPod:
        collectorName: "my-counter"
        timeout: 5s
        podSpec:
          containers:
          - name: my-counter
            image: ubuntu
            command: ["/bin/bash", "-c", 'for i in {1..30}; do echo "count $i"; sleep 1; done']
```

## Output
```
./bin/support-bundle test2.yaml -v0                                

 * failed to run collector: run-pod/my-counter: timeout
 Collecting support bundle ⠇ cluster-resources
support-bundle-2023-05-30T12_49_53.tar.gz

============ Collectors summary =============
Suceeded (S), eXcluded (X), Failed (F)
=============================================
run-pod/my-counter (F) : 34,059ms
cluster-resources (S)  : 1,877ms
cluster-info (S)       : 5ms

============ Redactors summary =============
In-cluster collectors : 375ms

============= Analyzers summary =============
Suceeded (S), eXcluded (X), Failed (F)
=============================================
No analyzers executed

Duration: 36,339ms
```

**Comment by DexterYan** [comment]
## Example Preflight
```
apiVersion: troubleshoot.sh/v1beta2
kind: Preflight
metadata:
  name: sample
spec:
  collectors:
    - run:
        collectorName: "static-hi"
        image: "alpine:3"
        command: ["echo", "hi static!"]
  analyzers:
    - textAnalyze:
        checkName: Said hi!
        fileName: /static-hi.log
        regex: "hi static"
        outcomes:
          - fail:
              message: Didn't say hi.
          - pass:
              message: Said hi!

```

## output
```
./bin/preflight test3.yaml -v0

============ Collectors summary =============
Suceeded (S), eXcluded (X), Failed (F)
=============================================
run/static-hi (S)     : 4,095ms
cluster-resources (S) : 1,877ms
cluster-info (S)      : 1ms

============ Redactors summary =============
No redactors executed

============= Analyzers summary =============
Suceeded (S), eXcluded (X), Failed (F)
=============================================
Said hi! (S) : 0ms

Duration: 7,779ms
```

**Comment by DexterYan** [comment]
## Previous Failed Support Bundle
```
apiVersion: troubleshoot.sh/v1beta2
kind: SupportBundle
metadata:
  name: example
spec:
  collectors:
    - runPod:
        collectorName: "my-counter"
        timeout: 5s
        podSpec:
          containers:
            - name: my-counter
              image: ubuntu
              command:
                [
                  "/bin/bash",
                  "-c",
                  'for i in {1..2}; do echo "count $i"; sleep 1; done',
                ]
    - runPod:
        collectorName: "my-counter"
        timeout: 5s
        podSpec:
          containers:
            - name: my-counter
              image: ubuntu
              command:
                [
                  "/bin/bash",
                  "-c",
                  'for i in {1..2}; do echo "count again $i"; sleep 1; done',
                ]

```
## Output
```
./bin/support-bundle test4.yaml -v0

 * failed to run collector: run-pod/my-counter: timeout
 * failed to run collector: run-pod/my-counter: timeout
 Collecting support bundle ⠏ cluster-info
support-bundle-2023-05-30T13_01_54.tar.gz

============ Collectors summary =============
Suceeded (S), eXcluded (X), Failed (F)
=============================================
run-pod/my-counter (F) : 7,044ms
cluster-resources (S)  : 1,875ms
cluster-info (S)       : 6ms

============ Redactors summary =============
In-cluster collectors : 386ms

============= Analyzers summary =============
Suceeded (S), eXcluded (X), Failed (F)
=============================================
No analyzers executed

Duration: 16,387ms
```

**Review by CpuID — APPROVED** [review]

File: pkg/collect/run_pod.go [hunk 1]
```diff
@@ -58,6 +59,26 @@ func (c *CollectRunPod) Collect(progressChan chan<- interface{}) (CollectorResul
 		if err := client.CoreV1().Pods(pod.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{}); err != nil {
 			klog.Errorf("Failed to delete pod %s: %v", pod.Name, err)
 		}
+		// Wait until the pod is deleted
+		// 2 minute for the maximum amount of time to wait for Pod deletion.
+		const gracePeriodMinutes = 2 * time.Minute
```
Comment (banjoh) [inline pkg/collect/run_pod.go:64]: Please create constant in `constants.go`

File: pkg/collect/run_pod.go [hunk 2]
```diff
@@ -58,6 +59,26 @@ func (c *CollectRunPod) Collect(progressChan chan<- interface{}) (CollectorResul
 		if err := client.CoreV1().Pods(pod.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{}); err != nil {
 			klog.Errorf("Failed to delete pod %s: %v", pod.Name, err)
 		}
+		// Wait until the pod is deleted
```
Comment (banjoh) [inline pkg/collect/run_pod.go:62]: Log here to let the user know that we are deleting the pod.

**Review by banjoh — CHANGES_REQUESTED** [review]
Please add logs (`klog.V(2)`) to show progress. This will come in handy whenever troubleshooting to see what the collector is doing. I suggest logging at least the following.
* Pod <name> has been created in namespace ...
* .... created in namespace with a timeout of <num>
* Pod <name> has been scheduled for deletion
* Polling pod <name> for deletion

The log lines can be whatever you see best.

**Comment by banjoh** [comment]
Overall, the approach you have here is good.

We would need to update [the documentation](https://troubleshoot.sh/docs/collect/run-pod/) to reflect this new changes. We also need to correct [the timeout](https://troubleshoot.sh/docs/collect/run-pod/#timeout-optional) documentation to reflect this change. Its not `30s` max, rather its `2mins + timeout` at the worst case, and if the parameter is not defined, the pod can potentially run forever.

**Comment by DexterYan** [comment]
Hey @banjoh 
Thank you for reviewing my PR. I have changed my pr based on your advice. And I think your first suggestion about using force pod deletion is necessary for solving issue #1188 

I have created another PR to reflect those changes 
#1196

**Pull request closed without merging**